File: js/packages/sdk/src/connections/connections.ts
import { create } from "@bufbuild/protobuf";
import { randomBytes } from "@noble/hashes/utils.js";
import {
  ConnectRequestSchema,
  ConnectResponseSchema,
  ConnectRotateSchema,
  ConnectCloseSchema,
  type ConnectRequest,
  type ConnectResponse,
  type ConnectRotate,
  type ConnectClose,
} from "../blockpb/block_types_pb.js";
import { encapsulate, decapsulate, makeKyberPair, type KeyPair } from "../crypto/index.js";
import { codeHex } from "../derive/index.js";
import type { Identity } from "../identity/index.js";
import { audienceCode, audienceSecret, connectionSecret0, rotatedSecret } from "./derive.js";

export type { ConnectRequest, ConnectResponse, ConnectRotate, ConnectClose };

/** Length of the handshake nonces. */
export const HANDSHAKE_NONCE_SIZE = 32;

export class WrongTargetError extends Error {
  constructor() {
    super("connections: request target is not this identity");
    this.name = "WrongTargetError";
  }
}
export class RotateMismatchError extends Error {
  constructor() {
    super("connections: rotation channel mismatch");
    this.name = "RotateMismatchError";
  }
}

/**
 * A remote identity addressable for a connection: its address and static KEM
 * public key (learned from its identity block).
 */
export interface Peer {
  readonly address: string;
  readonly kyberPublicKey: Uint8Array;
}

/**
 * An established connection at a given epoch. Its audience secret and code drive
 * the private channel's encryption.
 */
export class Connection {
  #epoch: number;
  #secret: Uint8Array;
  readonly #self: Identity;
  readonly #peerAddress: string;
  readonly #peerKyber: Uint8Array;

  constructor(self: Identity, peerAddress: string, peerKyber: Uint8Array, secret: Uint8Array) {
    this.#self = self;
    this.#peerAddress = peerAddress;
    this.#peerKyber = peerKyber;
    this.#epoch = 0;
    this.#secret = secret;
  }

  get epoch(): number {
    return this.#epoch;
  }

  get peerAddress(): string {
    return this.#peerAddress;
  }

  /** The current epoch's private audience secret. */
  audienceSecret(): Uint8Array {
    return audienceSecret(this.#secret, this.#epoch);
  }

  /** The current epoch's private audience code. */
  audienceCode(): Uint8Array {
    return audienceCode(this.#secret, this.#epoch);
  }

  /**
   * Advance this connection to the next epoch with fresh KEM entropy and return
   * the connect.rotate payload (to be sent encrypted to the current audience).
   */
  rotate(): ConnectRotate {
    const { ciphertext: ct3, sharedSecret: ss3 } = encapsulate(this.#peerKyber);
    this.#secret = rotatedSecret(this.#secret, ss3, ct3, this.#epoch + 1);
    this.#epoch += 1;
    return create(ConnectRotateSchema, {
      newChannel: codeHex(this.audienceCode()),
      kemCiphertext: ct3,
    });
  }

  /**
   * Advance this connection in response to a peer's connect.rotate, verifying
   * that the recomputed audience code matches new_channel before committing.
   */
  applyRotate(rot: ConnectRotate): void {
    const ss3 = decapsulate(rot.kemCiphertext, this.#self.kyber.secretKey);
    const next = rotatedSecret(this.#secret, ss3, rot.kemCiphertext, this.#epoch + 1);
    if (codeHex(audienceCode(next, this.#epoch + 1)) !== rot.newChannel) {
      throw new RotateMismatchError();
    }
    this.#secret = next;
    this.#epoch += 1;
  }
}

/**
 * The initiator's state between sending connect.request and receiving
 * connect.response.
 */
export class PendingRequest {
  constructor(
    private readonly self: Identity,
    private readonly peer: Peer,
    private readonly eph: KeyPair,
    private readonly ss1: Uint8Array,
    private readonly nonce: Uint8Array,
  ) {}

  /**
   * Finish the initiator's handshake from the connect.response, recovering the
   * second shared secret and deriving the Connection.
   */
  complete(resp: ConnectResponse): Connection {
    const ss2 = decapsulate(resp.kemCiphertext, this.eph.secretKey);
    const secret = connectionSecret0(this.ss1, ss2, this.nonce, resp.nonceResponse);
    return new Connection(this.self, this.peer.address, this.peer.kyberPublicKey, secret);
  }
}

/**
 * Build a connect.request to peer: a fresh ephemeral KEM key for forward
 * secrecy, a shared secret encapsulated to the peer's static key, and a random
 * nonce. Complete the returned PendingRequest once the response arrives.
 */
export function startRequest(
  self: Identity,
  peer: Peer,
): { pending: PendingRequest; request: ConnectRequest } {
  const eph = makeKyberPair(randomBytes(32));
  const { ciphertext: ct1, sharedSecret: ss1 } = encapsulate(peer.kyberPublicKey);
  const nonce = randomBytes(HANDSHAKE_NONCE_SIZE);
  const request = create(ConnectRequestSchema, {
    target: peer.address,
    initKemPub: eph.publicKey,
    kemCiphertext: ct1,
    nonce,
  });
  return { pending: new PendingRequest(self, peer, eph, ss1, nonce), request };
}

/**
 * Consume a connect.request addressed to self from initiator, returning the
 * connect.response payload and the established Connection. requestBlockID is the
 * block ID of the request (carried in the response for correlation).
 */
export function acceptRequest(
  self: Identity,
  initiator: Peer,
  request: ConnectRequest,
  requestBlockID: string,
): { response: ConnectResponse; connection: Connection } {
  if (request.target !== self.address) throw new WrongTargetError();
  const ss1 = decapsulate(request.kemCiphertext, self.kyber.secretKey);
  const { ciphertext: ct2, sharedSecret: ss2 } = encapsulate(request.initKemPub);
  const nonceResponse = randomBytes(HANDSHAKE_NONCE_SIZE);
  const secret = connectionSecret0(ss1, ss2, request.nonce, nonceResponse);
  const connection = new Connection(self, initiator.address, initiator.kyberPublicKey, secret);
  const response = create(ConnectResponseSchema, {
    request: requestBlockID,
    kemCiphertext: ct2,
    nonceResponse,
  });
  return { response, connection };
}

/** Build a connect.close payload notifying peer of teardown. */
export function buildClose(peerAddress: string, reason: string): ConnectClose {
  return create(ConnectCloseSchema, { target: peerAddress, reason });
}

/** A peer's static KEM public key as a Peer, given its identity. */
export function peerFromIdentity(id: Identity): Peer {
  return { address: id.address, kyberPublicKey: id.kyber.publicKey };
}

File: implementations/go/connections/connections.go
package connections

import (
	"crypto/rand"
	"errors"

	"github.com/cloudflare/circl/kem"

	"github.com/bpprotocol/blockparty/implementations/go/blockpb"
	"github.com/bpprotocol/blockparty/implementations/go/crypto"
	"github.com/bpprotocol/blockparty/implementations/go/derive"
	"github.com/bpprotocol/blockparty/implementations/go/identity"
)

// NonceSize is the length of the handshake nonces.
const NonceSize = 32

var (
	// ErrWrongTarget is returned when a request's target is not this identity.
	ErrWrongTarget = errors.New("connections: request target is not this identity")
	// ErrRotateMismatch is returned when a rotation's new_channel does not match
	// the recipient's recomputed audience code.
	ErrRotateMismatch = errors.New("connections: rotation channel mismatch")
)

// Peer is a remote identity addressable for a connection: its address and
// static KEM public key (learned from its identity block).
type Peer struct {
	Address  derive.Address
	KyberPub kem.PublicKey
}

// Connection is an established connection at a given epoch. Its audience secret
// and code drive the private channel's encryption (see the encryption package).
type Connection struct {
	self      identity.Identity
	peerAddr  derive.Address
	peerKyber kem.PublicKey
	epoch     uint64
	secret    []byte // connectionSecret(epoch)
}

// Epoch returns the connection's current epoch.
func (c *Connection) Epoch() uint64 { return c.epoch }

// PeerAddress returns the connected peer's address.
func (c *Connection) PeerAddress() derive.Address { return c.peerAddr }

// AudienceSecret returns the current epoch's private audience secret.
func (c *Connection) AudienceSecret() []byte { return audienceSecret(c.secret, c.epoch) }

// AudienceCode returns the current epoch's private audience code.
func (c *Connection) AudienceCode() derive.Code { return audienceCode(c.secret, c.epoch) }

// PendingRequest holds the initiator's state between sending connect.request
// and receiving connect.response.
type PendingRequest struct {
	self  identity.Identity
	peer  Peer
	eph   crypto.KyberKeyPair
	ss1   []byte
	nonce []byte
}

// StartRequest builds a connect.request to peer: a fresh ephemeral KEM key for
// forward secrecy, a shared secret encapsulated to the peer's static key, and a
// random nonce. The returned PendingRequest is completed with Complete once the
// response arrives.
func StartRequest(self identity.Identity, peer Peer) (*PendingRequest, *blockpb.ConnectRequest, error) {
	eph := crypto.MakeKyberPair(randomBytes(32))
	ct1, ss1, err := crypto.Encapsulate(peer.KyberPub)
	if err != nil {
		return nil, nil, err
	}
	nonce := randomBytes(NonceSize)
	req := &blockpb.ConnectRequest{
		Target:        string(peer.Address),
		InitKemPub:    eph.PublicBytes(),
		KemCiphertext: ct1,
		Nonce:         nonce,
	}
	return &PendingRequest{self: self, peer: peer, eph: eph, ss1: ss1, nonce: nonce}, req, nil
}

// AcceptRequest consumes a connect.request addressed to self from initiator,
// returning the connect.response payload and the established Connection.
// requestBlockID is the block ID of the request (carried in the response for
// correlation). The caller resolves the initiator's Peer (address + static KEM
// key) from the request's author before calling.
func AcceptRequest(self identity.Identity, initiator Peer, req *blockpb.ConnectRequest, requestBlockID string) (*blockpb.ConnectResponse, *Connection, error) {
	if req.Target != string(self.Address) {
		return nil, nil, ErrWrongTarget
	}
	ss1, err := self.Kyber.Decapsulate(req.KemCiphertext)
	if err != nil {
		return nil, nil, err
	}
	ephPub, err := crypto.KEMScheme().UnmarshalBinaryPublicKey(req.InitKemPub)
	if err != nil {
		return nil, nil, err
	}
	ct2, ss2, err := crypto.Encapsulate(ephPub)
	if err != nil {
		return nil, nil, err
	}
	nonceResp := randomBytes(NonceSize)
	conn := &Connection{
		self:      self,
		peerAddr:  initiator.Address,
		peerKyber: initiator.KyberPub,
		epoch:     0,
		secret:    connectionSecret0(ss1, ss2, req.Nonce, nonceResp),
	}
	resp := &blockpb.ConnectResponse{
		Request:       requestBlockID,
		KemCiphertext: ct2,
		NonceResponse: nonceResp,
	}
	return resp, conn, nil
}

// Complete finishes the initiator's handshake from the connect.response,
// recovering the second shared secret and deriving the Connection.
func (p *PendingRequest) Complete(resp *blockpb.ConnectResponse) (*Connection, error) {
	ss2, err := p.eph.Decapsulate(resp.KemCiphertext)
	if err != nil {
		return nil, err
	}
	return &Connection{
		self:      p.self,
		peerAddr:  p.peer.Address,
		peerKyber: p.peer.KyberPub,
		epoch:     0,
		secret:    connectionSecret0(p.ss1, ss2, p.nonce, resp.NonceResponse),
	}, nil
}

// Rotate advances this connection to the next epoch with fresh KEM entropy and
// returns the connect.rotate payload (to be sent encrypted to the current
// audience). The connection is mutated to the new epoch on success.
func (c *Connection) Rotate() (*blockpb.ConnectRotate, error) {
	ct3, ss3, err := crypto.Encapsulate(c.peerKyber)
	if err != nil {
		return nil, err
	}
	c.secret = rotatedSecret(c.secret, ss3, ct3, c.epoch+1)
	c.epoch++
	return &blockpb.ConnectRotate{NewChannel: c.AudienceCode().Hex(), KemCiphertext: ct3}, nil
}

// ApplyRotate advances this connection in response to a peer's connect.rotate,
// recovering the fresh entropy and verifying that the recomputed audience code
// matches the payload's new_channel before committing.
func (c *Connection) ApplyRotate(rot *blockpb.ConnectRotate) error {
	ss3, err := c.self.Kyber.Decapsulate(rot.KemCiphertext)
	if err != nil {
		return err
	}
	next := rotatedSecret(c.secret, ss3, rot.KemCiphertext, c.epoch+1)
	if audienceCode(next, c.epoch+1).Hex() != rot.NewChannel {
		return ErrRotateMismatch
	}
	c.secret = next
	c.epoch++
	return nil
}

// BuildClose builds a connect.close payload notifying peer of teardown.
func BuildClose(peer derive.Address, reason string) *blockpb.ConnectClose {
	return &blockpb.ConnectClose{Target: string(peer), Reason: reason}
}

func randomBytes(n int) []byte {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		panic("connections: out of randomness: " + err.Error())
	}
	return b
}
